stan/math/prim/mat/fun/gp_dot_prod_cov.hpp

```diff
@@ -0,0 +1,220 @@
+ * @return dot product kernel
+ * @throw std::domain_error if sigma < 0, nan, inf or
+ *   x is nan or infinite
+ */
+template <typename T_x, typename T_sigma>
+inline typename Eigen::Matrix<typename stan::return_type<T_x, T_sigma>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_dot_prod_cov(const std::vector<T_x> &x, const T_sigma &sigma) {
+  using stan::math::dot_product;
+  using stan::math::square;
+
+  check_not_nan("gp_dot_prod_cov", "sigma", sigma);
+  check_nonnegative("gp_dot_prod_cov", "sigma", sigma);
+  check_finite("gp_dot_prod_cov", "sigma", sigma);
+
+  for (size_t n = 0; n < x.size(); ++n) {
+    check_not_nan("gp_dot_prod_cov", "x", x[n]);
+    check_finite("gp_dot_prod_cov", "x", x[n]);
+  }
+
+  Eigen::Matrix<typename stan::return_type<T_x, T_sigma>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x.size(), x.size());
+  int x_size = x.size();
+  if (x_size == 0)
+    return cov;
+
+  T_sigma sigma_sq = square(sigma);
+
+  for (int i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq + dot_product(x[i], x[i]);
+    for (int j = i + 1; j < x_size; ++j) {
+      cov(i, j) = sigma_sq + dot_product(x[i], x[j]);
+      cov(j, i) = cov(i, j);
+    }
+  }
+  cov(x_size - 1, x_size - 1)
+      = sigma_sq + dot_product(x[x_size - 1], x[x_size - 1]);
+  return cov;
+}
+
+/**
+ * Returns a dot product kernel.
+ *
+ * @tparam T_x type of std::vector of double
+ * @tparam T_sigma type of sigma
+ *
+ * @param x std::vector of elements that can be used in transpose
+ *   and multiply
+ *    This function assumes each element of x is the same size.
+ * @param sigma
+ * @return dot product kernel
+ * @throw std::domain_error if sigma < 0, nan, inf or
+ *   x is nan or infinite
+ */
+template <typename T_sigma>
+inline typename Eigen::Matrix<typename stan::return_type<double, T_sigma>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_dot_prod_cov(const std::vector<double> &x, const T_sigma &sigma) {
+  using stan::math::dot_product;
+  using stan::math::square;
+
+  check_not_nan("gp_dot_prod_cov", "sigma", sigma);
+  check_nonnegative("gp_dot_prod_cov", "sigma", sigma);
+  check_finite("gp_dot_prod_cov", "sigma", sigma);
+
+  for (size_t n = 0; n < x.size(); ++n) {
+    check_not_nan("gp_dot_prod_cov", "x", x[n]);
+    check_finite("gp_dot_prod_cov", "x", x[n]);
+  }
+
+  Eigen::Matrix<typename stan::return_type<double, T_sigma>::type,
+                Eigen::Dynamic, Eigen::Dynamic>
+      cov(x.size(), x.size());
+  int x_size = x.size();
+  if (x_size == 0)
+    return cov;
+
+  T_sigma sigma_sq = square(sigma);
+
+  for (int i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq + x[i] * x[i];
+    for (int j = i + 1; j < x_size; ++j) {
+      cov(i, j) = sigma_sq + x[i] * x[j];
+      cov(j, i) = cov(i, j);
+    }
+  }
+  cov(x_size - 1, x_size - 1) = sigma_sq + x[x_size - 1] * x[x_size - 1];
+  return cov;
+}
+
+/**
+ * Returns a dot product kernel.
```

Comment:
Same comment as for first one---this needs some description and also some description of why there are two versions.